pkg/operator/webhookdeployment/webhook.go

```diff
@@ -0,0 +1,162 @@
+	client operatorclient.OperatorClient,
+	deployInformer appsinformersv1.DeploymentInformer,
+	kubeClient kubernetes.Interface,
+	versionGetter status.VersionGetter,
+	eventRecorder events.Recorder,
+	operandVersion string,
+	csiSnapshotWebhookImage string,
+) factory.Controller {
+	c := &csiSnapshotWebhookController{
+		client:                  client,
+		kubeClient:              kubeClient,
+		versionGetter:           versionGetter,
+		eventRecorder:           eventRecorder,
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-snapshot-controller"),
+		operandVersion:          operandVersion,
+		csiSnapshotWebhookImage: csiSnapshotWebhookImage,
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(client).WithInformers(
+		client.Informer(),
+		deployInformer.Informer(),
+	).ToController(WebhookControllerName, eventRecorder)
+}
+
+func (c *csiSnapshotWebhookController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, opStatus, _, err := c.client.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	deployment, err := c.getDeployment(opSpec)
+	if err != nil {
+		// This will set Degraded condition
+		return err
+	}
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(deployment, opStatus.Generations)
+	deployment, _, err = resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), c.eventRecorder, deployment, lastGeneration)
+	if err != nil {
+		// This will set Degraded condition
+		return err
+	}
+
+	// Compute status
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: WebhookControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "WaitDeployment"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: WebhookControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				c.versionGetter.SetVersion(webhookVersionName, c.operandVersion)
```

Comment:
In theory `Available` could still be `false` here. In that case, setting the operand version here would be incorrect.

Reply:
Also, we should only report `operator` version only after all operands have been rolled out: 

https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go#L45

However, currently we report the operator version in the other controller, which has no idea if the webhook operand has been rolled out yet.

How can we solve this? Another controller only to set the operator version? :thinking:

Reply:
It sets its own version, `webhookVersionName` - `versions` is an array. So there are two distinct version, which can should converge to the same value. Honestly, I don't know what CVO does with this array :-)

Reply:
Used Deploying everywhere

Reply:
I mean, I'm concerned about 2 things:

1. Setting `webhookVersionName` should be OK here, but I think we need to make sure we do so only when `Progressing=false` and `Available=true`. Currently it's setting when `Progressing=false`, which is too soon.

2. Since we're introducing a new operand (the webhook), we need to change this part of the code to only set the `operator` when both operands (webhook and the snapshot-controller) have rolled out (currently it only accounts for the snapshot-controller):

https://github.com/openshift/cluster-csi-snapshot-controller-operator/blob/master/pkg/operator/sync.go#L170

Reply:
I completely removed the version, I need to think about it more.